src/languages/swift.js

```diff
@@ -69,12 +69,30 @@ export default function(hljs) {
       {begin: /"/, end: /"/},
     ]
   };
-  var NUMBERS = {
+
+  // https://docs.swift.org/swift-book/ReferenceManual/LexicalStructure.html#grammar_numeric-literal
+  var decimalDigits = '([0-9]_*)+';
+  var hexDigits = '([0-9a-fA-F]_*)+';
+  var NUMBER = {
       className: 'number',
-      begin: '\\b([\\d_]+(\\.[\\deE_]+)?|0x[a-fA-F0-9_]+(\\.[a-fA-F0-9p_]+)?|0b[01_]+|0o[0-7_]+)\\b',
-      relevance: 0
+      relevance: 0,
+      variants: [
+        // decimal floating-point-literal (subsumes decimal-literal)
+        { begin: `-?\\b(${decimalDigits})(\\.(${decimalDigits}))?` +
```

Comment:
Has this not come up before or have I missed this on every other PR?

How does this know the difference between a unary - and a - operator?

```
a = -3
// vs
5-3
```

The `-` in the latter can't be part of the number without a proper full parser.

Reply:
In the other languages, leading `-` is never part of the number (it's just a preceding unary operator)—but that's not the case with Swift. You're right that this won't be fully solved, so it's necessary instead to pick a "good enough" behavior. How do you feel about lookbehind? Without using it, I think this is the best tradeoff, but with it, there could be an assertion that the leading `-` is not preceded by a literal or identifier character (e.g., something like ```((?<!"#*|[\\w\\xA8-\\uFFFD]\`?)-)?\\b(${decimalDigits})…```, assuming these are non-Unicode regular expressions).

Reply:
Can't use look-behind, it's not supported on Safari yet.  There are a few hacks (fake look-behind, callbacks, etc.) we could use but in other grammars we just take the expedient route and just do not include `-` as part of the number literal. 

> In the other languages, leading - is never part of the numeric literal (it's just a preceding unary operator)—but that's not the case with Swift. 

I'm not sure this is relevant for us.  

> Without using it, I think this is the best tradeoff

This will always mis-highlight `3-2` (which is bad and just looks weird) and it's _inconsistent_ with the way we handle this is every other grammar.  Plus `-` should really (in most cases) be highlighted as an operator (with our higher fidelity initiative).  _For things like this the core library strives for overall consistency among grammars rather than absolute adherence to the quirks of any particular language._  Highlighting a snippet of JavaScript vs a snippet of Swift side by side should not result in different behavior for unary literals.

So for consistency let's drop the `-?` from all the rules and you can include a TODO note about the behavior and mention lookbehind - and this matter can be revisited in the future when we have better tools at our disposal.

Reply:
OK, that makes sense.